### ISSUE TITLE
Allow role templates / inheritance

### ISSUE THREAD

**Issue by brianjrobertson** [holacracyone/Holacracy-Constitution#58]
As the Lead Link of one of many sub-circles that do a very similar type of work with similar roles, I'd like to be able to start my circle with a set of default/templated roles defined and maintained elsewhere in the organization, and then modify them only as-needed to adapt to our local context, so that I can benefit from prior organizational wisdom/learning without reinventing the wheel in my sub-circle.

**Comment by brianjrobertson** [comment]
The implementation vision I'm currently playing with here is to allow roles to "inherit" from other roles in other circles.  So, for example, a project-team sub-circle could have some common roles that inherit from templates in the super-circle.  Instead of creating its own "Project Manager", "Developer", and "Designer" roles that look similar to the ones in other sub-circles, it could just have those roles inherit the purpose and accountabilities of the default/template version of that role in the super-circle.  Then, when learning happens that should affect _all_ Developers, someone can propose adding an accountability on the template role in the super-circle (and all sub-circle Developers that inherit from it will automatically gain that accountability).  When learning happens that's local to one project team, that project team would simply update its own version of the role, and thus add to or modify the inherited accountabilities/purpose.

I do see some issues and shortcomings of this implementation strategy.  As one example, it'd be nice to be able to have a whole type of sub-circle that inherits an entire default set of roles, rather than just role-level inheritance, so the sub-circle doesn't even need to create those roles individually.  But I'm not quite sure how that would/could work.

As another example of a potential issue here, this approach creates a weirdness of how those template roles are filled/enacted in the super-circle; they could be meant just as templates for sub-circles to inherit from, and not actually as operating entities in the super-circle, in which case it's very weird that they'd be filled by someone (even if Lead Link by default).  We could address that by effectively allowing a role to be created purely as a "virtual" or "template" role that's only there to be inherited from, and exclude that from all operational requirements/duties.  That would solve it, and offers some interesting possibilities for structuring, but it also adds some complexity to the basic rules.

Any reactions (especially better ideas)?

**Comment by djwork** [comment]
Quick thoughts....
I would:
  leave inheritance out of the constitution
  Streamline the creation of holacracy apps
  Declare the template roles in an app
  Declare the inheritance rules in the app policy
  When a circle through governance implements the app inheritance rule for the instance roles may override the inheritance policy of the app.
  The app should also define roles/policy for update of the template roles in the app.

In short for version 5 I would keep inheritance out of the OS and trial it in the APP layer

**Comment by dlocke** [comment]
Brian,

This seems like this could easily be done with the current constitution.  Am I missing something?

David

**Comment by tensiondriven** [comment]
I feel that adding inheritance to the constitution would make it much harder to understand, as it adds a level of abstraction that people aren't accustom to thinking about.  I agree with @djwork - it seems that the current constitution allows for this.

My reaction is that this seems like a useful feature of a holacracy-management tool like Glassfrog than an enhancement to the constitution.

**Comment by brianjrobertson** [comment]
Thanks folks, that was all very helpful.  I think a next-step here might be to try writing the roles/policies needed to effect this as an app, and see if/where it's unclear whether those policies are constitutional (v4.1 isn't very clear on what policies you can or can't adopt when they skirt the line around stuff that's normally done via governance)...

**Comment by cassus** [comment]
I like the idea of having this feature in app-space. Still there will be one implementation of this in GlassFrog and it will be painful to deviate from it. In practice the features in GlassFrog will dictate how we do template roles, analogue to the Async governance.

My 2 cents it to work on the role-template app in tandem with the UX design of the GlassFrog "reference-implementation". Kind of do user-centered design on the app itself.

**Comment by andreafare** [comment]
I like the template idea, recently I've been thinking about  specific implementation patterns that would benefit from templates. 

Sticking with the software lingo, it would be nice to have some sort of a "templated role factory", where you could start a new circle picking instantiations from some sort of a predefined role template catalogue (an abstract role repository). It would just become a predefined way to start a new circle with the option of composing its starting lego bricks, then Governance would take off and those roles would be modified as required by the specific needs of each circle.

This specific requirement doesn't even need to be an app, it could be just a new feature in Glassfrog.

I too would leave inheritance out of the game at least at runtime, it would make the dynamics of Circles way more complex in my opinion...

**Comment by julianeroell** [comment]
So far, we have not seen the need for inheritance / templates in our organisation. I am with the previous posters here in that I'd prefer to see this as an app or as an implementation in Glassfrog, rather than as an addition to the constitution.

**Comment by KoenVeltman** [comment]
(withouth having read all the responses) adding my 5 cents. it would be great just to have the freedom to copy existing roles/policies. I have seen the need for this in a few organizations. Especially when an organization is "project" or "process" based. Examples I have seen include a consultancy, some manufacturers or geographically organized with similar organizations in each geography. Most notably I visited a nursery once which had this need, 50 self-organized locations with similar "operations".

It would be great to just copy an existing circle and use that as a starting point to evolve and adapt then locally. (vs the work around to just manually copy all roles/accountabilities.

**Comment by lmmathurin** [comment]
Like others comments.  If I understand the User Story.  It refer to **start** or initial roles in the sub-circle.  So even in the 4.1 constitution this is "possible".  Let me explain, this is just a feature needed in an Holacracy application(like Glassfrog) to copy roles / policies to create an **initial proposal**.  No governance need at this point.

With this facility, everything is crafted in advance for the first Governance Meeting, or can be process asynchronously!  So technically, since this is a brand new Sub-Circle without any roles inside of it except Core Roles, only the Lead Link of this Sub-Circle can ask for integrating in governance...if process asynchronously. 

So, I don't see the needs for changing the constitution with this user story.

If you'd like to discuss about Inheritance,  this is not a good design pattern when programming in my opinion.  In the case you only add accountabilities to the inherited ones this is fast, but can be far more complex if you can override an accountability of the inherited role.  Composition pattern is usually prefer: https://en.wikipedia.org/wiki/Composition_over_inheritance.

**Comment by brianjrobertson** [comment]
Dropping this as unneeded.
